File: analyze_microlessons.py
#!/usr/bin/env python3
import yaml
import os
from pathlib import Path
from collections import defaultdict

# Directory containing microlesson files
LESSONS_DIR = "/home/user/idlecampus-backend/db/seeds/security_complete_enhanced/microlessons"

# Track issues
issues = defaultdict(list)
stats = {
    'total_files': 0,
    'files_with_exercises': 0,
    'files_without_exercises': 0,
    'mcq_with_placeholder_options': 0,
    'mcq_without_question': 0,
    'mcq_without_explanation': 0,
    'generic_exercises': 0,
    'terminal_exercises_incomplete': 0,
    'sandbox_exercises_incomplete': 0,
    'missing_content': 0,
    'yaml_errors': 0
}

def analyze_lesson(file_path):
    """Analyze a single lesson file"""
    filename = os.path.basename(file_path)

    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            data = yaml.safe_load(f)

        stats['total_files'] += 1
        file_issues = []

        # Check basic structure
        if not data:
            stats['yaml_errors'] += 1
            file_issues.append("Empty or invalid YAML")
            issues['yaml_errors'].append(filename)
            return file_issues

        # Check required fields
        required_fields = ['slug', 'title', 'content_md', 'exercises']
        missing_fields = [f for f in required_fields if f not in data]
        if missing_fields:
            file_issues.append(f"Missing fields: {', '.join(missing_fields)}")

        # Check content quality
        if 'content_md' in data:
            content = data['content_md']
            if not content or len(content.strip()) < 100:
                stats['missing_content'] += 1
                file_issues.append("Content is missing or too short")

        # Check exercises
        if 'exercises' not in data or not data['exercises']:
            stats['files_without_exercises'] += 1
            file_issues.append("No exercises found")
            issues['no_exercises'].append(filename)
        else:
            stats['files_with_exercises'] += 1
            exercises = data['exercises']

            for i, exercise in enumerate(exercises):
                ex_type = exercise.get('type', 'unknown')

                # Check MCQ exercises
                if ex_type == 'mcq':
                    mcq_issues = []

                    # Check for placeholder options (A, B, C, D)
                    options = exercise.get('options', [])
                    if options == ['A', 'B', 'C', 'D']:
                        stats['mcq_with_placeholder_options'] += 1
                        mcq_issues.append("Placeholder options (A, B, C, D)")
                        issues['placeholder_mcq'].append(filename)

                    # Check for missing question
                    question = exercise.get('question', '')
                    if not question or question == "Which statement is correct for this topic?":
                        stats['mcq_without_question'] += 1
                        mcq_issues.append("Generic/missing question")
                        issues['generic_mcq_question'].append(filename)

                    # Check for generic explanation
                    explanation = exercise.get('explanation', '')
                    if explanation == "Revisit the definitions and tool outputs for accuracy.":
                        stats['mcq_without_explanation'] += 1
                        mcq_issues.append("Generic explanation")
                        issues['generic_explanation'].append(filename)

                    # Check for missing options
                    if not options:
                        mcq_issues.append("No options provided")

                    # Check for correct answer
                    if 'correct_answer_index' not in exercise:
                        mcq_issues.append("No correct answer specified")

                    if mcq_issues:
                        file_issues.append(f"MCQ #{i+1}: {', '.join(mcq_issues)}")

                # Check terminal exercises
                elif ex_type == 'terminal':
                    if 'description' not in exercise or not exercise['description']:
                        stats['terminal_exercises_incomplete'] += 1
                        file_issues.append(f"Terminal exercise #{i+1}: Missing description")
                        issues['incomplete_terminal'].append(filename)

                # Check sandbox exercises
                elif ex_type == 'sandbox':
                    # Sandbox exercises might be minimal, but track them
                    if len(exercise.keys()) <= 1:  # Only has 'type'
                        stats['sandbox_exercises_incomplete'] += 1
                        file_issues.append(f"Sandbox exercise #{i+1}: Minimal/incomplete")
                        issues['incomplete_sandbox'].append(filename)

                # Check reflection exercises
                elif ex_type == 'reflection':
                    prompt = exercise.get('prompt', '')
                    if prompt == "Where does this show up in real incidents or reviews? What would you check first?":
                        stats['generic_exercises'] += 1
                        file_issues.append(f"Reflection #{i+1}: Generic prompt")
                        issues['generic_reflection'].append(filename)

                # Check checkpoint exercises
                elif ex_type == 'checkpoint':
                    prompt = exercise.get('prompt', '')
                    if prompt == "Write the exact steps/commands/config needed to apply this concept.":
                        stats['generic_exercises'] += 1
                        file_issues.append(f"Checkpoint #{i+1}: Generic prompt")
                        issues['generic_checkpoint'].append(filename)

        return file_issues

    except yaml.YAMLError as e:
        stats['yaml_errors'] += 1
        issues['yaml_errors'].append(filename)
        return [f"YAML parsing error: {str(e)}"]
    except Exception as e:
        return [f"Error: {str(e)}"]

def main():
    """Main analysis function"""
    all_files_issues = {}

    # Get all YAML files
    yaml_files = sorted(Path(LESSONS_DIR).glob('*.yml'))

    print(f"Analyzing {len(yaml_files)} microlesson files...\n")

    # Analyze each file
    for file_path in yaml_files:
        file_issues = analyze_lesson(str(file_path))
        if file_issues:
            all_files_issues[file_path.name] = file_issues

    # Print summary statistics
    print("=" * 80)
    print("MICROLESSON REVIEW REPORT")
    print("=" * 80)
    print()

    print("OVERALL STATISTICS:")
    print(f"  Total files reviewed: {stats['total_files']}")
    print(f"  Files with exercises: {stats['files_with_exercises']}")
    print(f"  Files without exercises: {stats['files_without_exercises']}")
    print()

    print("EXERCISE QUALITY ISSUES:")
    print(f"  MCQ with placeholder options (A,B,C,D): {stats['mcq_with_placeholder_options']}")
    print(f"  MCQ with generic/missing questions: {stats['mcq_without_question']}")
    print(f"  MCQ with generic explanations: {stats['mcq_without_explanation']}")
    print(f"  Generic reflection/checkpoint prompts: {stats['generic_exercises']}")
    print(f"  Incomplete terminal exercises: {stats['terminal_exercises_incomplete']}")
    print(f"  Incomplete sandbox exercises: {stats['sandbox_exercises_incomplete']}")
    print()

    print("CONTENT ISSUES:")
    print(f"  Missing or insufficient content: {stats['missing_content']}")
    print(f"  YAML parsing errors: {stats['yaml_errors']}")
    print()

    # Print pattern summary
    print("=" * 80)
    print("PATTERN ANALYSIS:")
    print("=" * 80)
    print()

    for issue_type, files in sorted(issues.items()):
        if files:
            print(f"{issue_type.upper().replace('_', ' ')} ({len(files)} files)")
    print()

    # Print sample of files with issues
    print("=" * 80)
    print("SAMPLE FILES WITH ISSUES (First 15):")
    print("=" * 80)
    print()

    count = 0
    for filename, file_issues in sorted(all_files_issues.items()):
        if count >= 15:
            break
        print(f"\n{filename}:")
        for issue in file_issues:
            print(f"  - {issue}")
        count += 1

    if len(all_files_issues) > 15:
        print(f"\n... and {len(all_files_issues) - 15} more files with issues")

    # Estimate files needing fixes
    print()
    print("=" * 80)
    print("REPAIR ESTIMATE:")
    print("=" * 80)
    print()

    files_needing_fixes = len(all_files_issues)
    print(f"Files needing fixes: {files_needing_fixes} out of {stats['total_files']} ({files_needing_fixes/stats['total_files']*100:.1f}%)")

    # Categorize severity
    critical = []
    moderate = []
    minor = []

    for filename, file_issues in all_files_issues.items():
        issue_count = len(file_issues)
        has_placeholder = any('Placeholder options' in issue for issue in file_issues)
        has_no_exercises = any('No exercises' in issue for issue in file_issues)

        if has_no_exercises or issue_count >= 5:
            critical.append(filename)
        elif has_placeholder or issue_count >= 3:
            moderate.append(filename)
        else:
            minor.append(filename)

    print(f"\nSeverity breakdown:")
    print(f"  Critical (no exercises or 5+ issues): {len(critical)}")
    print(f"  Moderate (placeholder MCQs or 3-4 issues): {len(moderate)}")
    print(f"  Minor (1-2 issues): {len(minor)}")

    print()
    print("=" * 80)
    print("RECOMMENDATIONS:")
    print("=" * 80)
    print()
    print("1. Replace all placeholder MCQ options (A,B,C,D) with real, content-specific options")
    print("2. Create specific, meaningful questions for each MCQ based on lesson content")
    print("3. Write detailed explanations for each MCQ answer")
    print("4. Customize reflection and checkpoint prompts to match lesson topics")
    print("5. Complete terminal exercise descriptions with specific tasks")
    print("6. Review and enhance sandbox exercises with clear objectives")
    print("7. Ensure all exercises test concepts actually covered in the lesson content")
    print()

if __name__ == '__main__':
    main()

File: analyze_microlessons.rb
#!/usr/bin/env ruby
require 'yaml'
require 'json'

class MicrolessonAnalyzer
  attr_reader :issues, :stats

  def initialize
    @issues = []
    @stats = {
      total_files: 0,
      total_courses: 0,
      files_with_issues: 0,
      empty_content: 0,
      no_exercises: 0,
      malformed_yaml: 0,
      short_content: 0,
      exercise_issues: 0,
      missing_fields: 0
    }
  end

  def analyze_all
    base_path = 'db/seeds/converted_microlessons'
    courses = Dir.glob("#{base_path}/*").select { |f| File.directory?(f) }

    @stats[:total_courses] = courses.length

    courses.each do |course_dir|
      course_name = File.basename(course_dir)
      microlessons_dir = File.join(course_dir, 'microlessons')

      next unless Dir.exist?(microlessons_dir)

      Dir.glob("#{microlessons_dir}/*.yml").each do |file_path|
        @stats[:total_files] += 1
        analyze_file(file_path, course_name)
      end
    end

    generate_report
  end

  def analyze_file(file_path, course_name)
    file_name = File.basename(file_path)
    file_issues = []

    begin
      content = File.read(file_path)
      data = YAML.load(content)

      # Check for missing or empty content
      if data['content_md'].nil? || data['content_md'].to_s.strip.empty?
        file_issues << "Empty or missing content_md"
        @stats[:empty_content] += 1
      elsif data['content_md'].to_s.strip.length < 100
        file_issues << "Very short content (#{data['content_md'].to_s.strip.length} chars)"
        @stats[:short_content] += 1
      end

      # Check for missing exercises
      if data['exercises'].nil? || data['exercises'].empty?
        file_issues << "No exercises defined"
        @stats[:no_exercises] += 1
      else
        # Analyze exercises
        exercise_issues = analyze_exercises(data['exercises'])
        file_issues.concat(exercise_issues) unless exercise_issues.empty?
      end

      # Check for missing required fields
      required_fields = ['slug', 'title', 'difficulty', 'estimated_minutes']
      missing = required_fields.select { |f| data[f].nil? || data[f].to_s.strip.empty? }
      unless missing.empty?
        file_issues << "Missing required fields: #{missing.join(', ')}"
        @stats[:missing_fields] += 1
      end

      # Check for placeholder or generic content
      if data['content_md'] && data['content_md'].match?(/TODO|PLACEHOLDER|FIX ?ME|TBD/i)
        file_issues << "Contains placeholder text (TODO/PLACEHOLDER/FIXME/TBD)"
      end

    rescue Psych::SyntaxError => e
      file_issues << "YAML syntax error: #{e.message}"
      @stats[:malformed_yaml] += 1
    rescue => e
      file_issues << "Error reading file: #{e.message}"
    end

    unless file_issues.empty?
      @stats[:files_with_issues] += 1
      @issues << {
        course: course_name,
        file: file_name,
        path: file_path,
        issues: file_issues
      }
    end
  end

  def analyze_exercises(exercises)
    issues = []

    exercises.each_with_index do |exercise, idx|
      exercise_num = idx + 1

      # Check exercise type
      unless exercise['type']
        issues << "Exercise #{exercise_num}: Missing 'type' field"
        @stats[:exercise_issues] += 1
        next
      end

      case exercise['type']
      when 'mcq'
        if !exercise['question'] || exercise['question'].to_s.strip.empty?
          issues << "Exercise #{exercise_num} (MCQ): Missing question"
          @stats[:exercise_issues] += 1
        end

        if !exercise['options'] || exercise['options'].empty?
          issues << "Exercise #{exercise_num} (MCQ): Missing or empty options"
          @stats[:exercise_issues] += 1
        elsif exercise['options'].length < 2
          issues << "Exercise #{exercise_num} (MCQ): Less than 2 options"
          @stats[:exercise_issues] += 1
        end

        if !exercise['correct_answer']
          issues << "Exercise #{exercise_num} (MCQ): Missing correct_answer"
          @stats[:exercise_issues] += 1
        end

      when 'code', 'terminal'
        if !exercise['problem_statement'] || exercise['problem_statement'].to_s.strip.empty?
          issues << "Exercise #{exercise_num} (#{exercise['type']}): Missing problem_statement"
          @stats[:exercise_issues] += 1
        end

        if exercise['type'] == 'code'
          if !exercise['test_cases'] || exercise['test_cases'].empty?
            issues << "Exercise #{exercise_num} (code): Missing test_cases"
            @stats[:exercise_issues] += 1
          end

          if !exercise['language']
            issues << "Exercise #{exercise_num} (code): Missing language"
            @stats[:exercise_issues] += 1
          end
        end

      when 'short_answer'
        if !exercise['question'] || exercise['question'].to_s.strip.empty?
          issues << "Exercise #{exercise_num} (short_answer): Missing question"
          @stats[:exercise_issues] += 1
        end

        if !exercise['expected_answer'] || exercise['expected_answer'].to_s.strip.empty?
          issues << "Exercise #{exercise_num} (short_answer): Missing expected_answer"
          @stats[:exercise_issues] += 1
        end

      when 'sandbox'
        if !exercise['description'] || exercise['description'].to_s.strip.empty?
          issues << "Exercise #{exercise_num} (sandbox): Missing description"
          @stats[:exercise_issues] += 1
        end
      end
    end

    issues
  end

  def generate_report
    puts "\n" + "="*80
    puts "MICROLESSON REVIEW REPORT"
    puts "="*80

    puts "\n📊 OVERALL STATISTICS:"
    puts "-" * 80
    puts "Total Courses: #{@stats[:total_courses]}"
    puts "Total Microlesson Files: #{@stats[:total_files]}"
    puts "Files with Issues: #{@stats[:files_with_issues]} (#{percentage(@stats[:files_with_issues], @stats[:total_files])}%)"
    puts "Files without Issues: #{@stats[:total_files] - @stats[:files_with_issues]} (#{percentage(@stats[:total_files] - @stats[:files_with_issues], @stats[:total_files])}%)"

    puts "\n🔍 ISSUE BREAKDOWN:"
    puts "-" * 80
    puts "Empty or Missing Content: #{@stats[:empty_content]}"
    puts "Very Short Content (<100 chars): #{@stats[:short_content]}"
    puts "No Exercises: #{@stats[:no_exercises]}"
    puts "Exercise-related Issues: #{@stats[:exercise_issues]}"
    puts "Missing Required Fields: #{@stats[:missing_fields]}"
    puts "Malformed YAML: #{@stats[:malformed_yaml]}"

    # Group issues by course
    issues_by_course = @issues.group_by { |i| i[:course] }

    puts "\n📚 ISSUES BY COURSE:"
    puts "-" * 80
    issues_by_course.sort_by { |course, issues| -issues.length }.each do |course, course_issues|
      puts "\n#{course} (#{course_issues.length} files with issues):"
      course_issues.first(5).each do |issue|
        puts "  • #{issue[:file]}"
        issue[:issues].each do |msg|
          puts "    - #{msg}"
        end
      end
      if course_issues.length > 5
        puts "  ... and #{course_issues.length - 5} more files"
      end
    end

    # Top 10 most problematic files
    puts "\n⚠️  TOP 20 MOST PROBLEMATIC FILES:"
    puts "-" * 80
    top_issues = @issues.sort_by { |i| -i[:issues].length }.first(20)
    top_issues.each_with_index do |issue, idx|
      puts "\n#{idx + 1}. #{issue[:course]}/#{issue[:file]}"
      puts "   Path: #{issue[:path]}"
      puts "   Issues (#{issue[:issues].length}):"
      issue[:issues].each do |msg|
        puts "     • #{msg}"
      end
    end

    puts "\n" + "="*80
    puts "END OF REPORT"
    puts "="*80
  end

  private

  def percentage(part, whole)
    return 0 if whole == 0
    ((part.to_f / whole) * 100).round(1)
  end
end

# Run the analyzer
analyzer = MicrolessonAnalyzer.new
analyzer.analyze_all
